### PR TITLE
Avoid duplicate Vulkan layer

### PR DESCRIPTION
**Pull request by rschiu**
Bug: b/147832494

### CODE REVIEW

**Comment by lpy** [comment]
Should we do it in https://github.com/google/gapid/blob/master/core/os/android/layers.go#L47?

File: gapis/trace/tracer/tracer.go [hunk 1]
```diff
@@ -87,7 +87,16 @@ func LayersFromOptions(ctx context.Context, o *service.TraceOptions) []string {
 	for _, x := range o.PerfettoConfig.GetDataSources() {
 		if layer, err := layout.LayerFromDataSource(x.GetConfig().GetName()); err == nil {
 			if _, err := layout.LibraryFromLayerName(layer); err == nil {
-				ret = append(ret, layer)
+				skip := false
```
Comment (AWoloszyn) [inline gapis/trace/tracer/tracer.go]: I think this is correct, but can I recommend
```
Outer:
  for _, x := range o.PerfettoConfig.GetDataSources() {
    if layer, err := layout.LayerFromDataSource(x.GetConfig().GetName()); err == nil {
	 if _, err := layout.LibraryFromLayerName(layer); err == nil {
            for _, l := range ret {
	        if l == layer {
		   continue Outer
		}
             }
             ret = append(ret, layer)
          }
    }
  }
```

**Comment by AWoloszyn** [comment]
> Should we do it in https://github.com/google/gapid/blob/master/core/os/android/layers.go#L47?

I think this is the correct location, there is no situation in which we would want the same perfetto layer to get added twice.

File: gapis/trace/tracer/tracer.go (hunk 1 above)
Comment (rschiu) [reply gapis/trace/tracer/tracer.go]: Done.  I did not know that was an option!

File: gapis/trace/tracer/tracer.go [hunk 2]
```diff
@@ -84,9 +84,15 @@ func LayersFromOptions(ctx context.Context, o *service.TraceOptions) []string {
...
 	}
+Outer:
 	for _, x := range o.PerfettoConfig.GetDataSources() {
 		if layer, err := layout.LayerFromDataSource(x.GetConfig().GetName()); err == nil {
 			if _, err := layout.LibraryFromLayerName(layer); err == nil {
+				for _, l := range ret {
+					if l == layer {
+						continue Outer
+					}
+				}
```
Comment (pmuetschard) [inline gapis/trace/tracer/tracer.go]: Maube put them in a map and then return the map's keys? I guess N^2 here is ok, but...

Comment (rschiu) [reply gapis/trace/tracer/tracer.go]: I figure with the typical number of layers we have, the map might make it worse.  But I don't mind changing to map either.

Comment (rschiu) [reply gapis/trace/tracer/tracer.go]: Done

**Comment by rschiu** [comment]
@pmuetschard PTAL.  Thanks.

File: gapis/trace/tracer/tracer.go [hunk 3]
```diff
@@ -84,9 +84,15 @@ func LayersFromOptions(ctx context.Context, o *service.TraceOptions) []string {
...
 	}
+
+	added_layer := make(map[string]bool)
 	for _, x := range o.PerfettoConfig.GetDataSources() {
 		if layer, err := layout.LayerFromDataSource(x.GetConfig().GetName()); err == nil {
 			if _, err := layout.LibraryFromLayerName(layer); err == nil {
+				if added_layer[layer] {
+					continue
+				}
+				added_layer[layer] = true
```
Comment (pmuetschard) [inline gapis/trace/tracer/tracer.go]: This is OK, but the more, sigh, idiomatic go way to do this is:
```
added := map[string]struct{}{}
for _, x := ... {
  ...
  if _, ok := added[layer]; !ok {
    ret = append(ret, layer)
    added[layer] = struct{}{}
  }
}
```
(also, use camelCase not snake_case)

Comment (rschiu) [reply gapis/trace/tracer/tracer.go]: Thanks! Done.

**Pull request merged**